### PR TITLE
Fix of `.merge_single_cells()` to Load Single-Cell Data into Dataframes

### PR DESCRIPTION
**Pull request by bunnech**
# Description

This pull request addresses issue [#195](https://github.com/cytomining/pycytominer/issues/195), concerning a potential memory leak in the `.merge_single_cells()` function as mentioned in pull request [#194](https://github.com/cytomining/pycytominer/pull/194#issuecomment-1179340277) and issue [#215](https://github.com/cytomining/pycytominer/issues/215#issuecomment-1192028633).

The updated function `.merge_single_cells()` combined different compartments, image data, and metadata and loads them into a Pandas dataframes. The old version did not finish converting a >10 Gb `.sqlite` files into dataframes within 4 hours. The current fix finishes the task in < 15 minutes.
This was achieved by removing the dependency to `pd.read_sql` (as suggested by @johnarevalo) and by getting rid of temporarily created dataframes of compartments to be merged which took additional memory. There might be better solutions, but this is working well for me currently. 

Let me know what you think!

## What is the nature of your change?

- [x] Bug fix (fixes an issue).
- [ ] Enhancement (adds functionality).
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [ ] This change requires a documentation update.

# Checklist

Please ensure that all boxes are checked before indicating that a pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have deleted all non-relevant text in this pull request template.

### CODE REVIEW

File: pycytominer/cyto_utils/cells.py
Comment (gwaybio) [inline pycytominer/cyto_utils/cells.py:408]: please use a more descriptive name.

Maybe something like `count_sql_table_rows()`

File: pycytominer/cyto_utils/cells.py
Comment (gwaybio) [inline pycytominer/cyto_utils/cells.py:413]: more descriptive name please.

For this, how about `get_sql_table_col_names()`?

File: pycytominer/cyto_utils/cells.py
Comment (gwaybio) [inline pycytominer/cyto_utils/cells.py:455]: can you name `resultset` more descriptively?

File: pycytominer/cyto_utils/cells.py
Comment (gwaybio) [inline pycytominer/cyto_utils/cells.py:460]: Is there any reason to not include the row _and_ column indexing here?

(e.g. `metas.loc[i, :] = row[:num_meta=]`, and something equivalent for `feats`)

To me, it's easier to understand that the variables you're building are 2d if we explicitly set both indices. 

We shouldn't implement my suggestion if it slows processing

File: pycytominer/cyto_utils/cells.py
Comment (gwaybio) [inline pycytominer/cyto_utils/cells.py:463]: metadata should be indexed first, right? Am I reading correctly that it is indexed _after_ morphology features in this setup?

File: pycytominer/cyto_utils/cells.py
Comment (gwaybio) [inline pycytominer/cyto_utils/cells.py:710]: nice!!

File: pycytominer/tests/test_cyto_utils/test_cells.py
Comment (gwaybio) [inline pycytominer/tests/test_cyto_utils/test_cells.py:243]: I am surprised that this passes the tests, given this function:

```python
def is_feature_col(self, col):
    """Check if column is a feature."""
    return (
        col.startswith("Cell")
        or col.startswith("Cytoplasm")
        or col.startswith("Nuclei")
    )
```

Is it possible that all features prefixed with `new` are considered metadata features in your implementation?

File: pycytominer/tests/test_cyto_utils/test_cells.py
Comment (gwaybio) [inline pycytominer/tests/test_cyto_utils/test_cells.py:312]: This is a bit concerning to me. Can you report what changed and why ignoring column dtype is now necessary?

File: pycytominer/cyto_utils/cells.py
Comment (gwaybio) [inline pycytominer/cyto_utils/cells.py:400]: I made a comment about this function in the proposed test update.

I think that you can easily update this function to iterate through `self.compartments` rather than hardcoding canonical compartments here. (i.e. what about non-canonical compartments?)

**Review by gwaybio — CHANGES_REQUESTED** [review]
Thanks for this contribution! I made several comments throughout that we should discuss prior to merging.

I have two additional discussion items as well. Can you comment on:

1. I'm wondering why we need to separate metadata features and morphology features at all. Would it make sense to process all features the same way? Ah, but maybe if we do that they won't be in the correct order in the final sc_df?
2. I am inclined to ask for tests for the three new functions you added (prior to name change, `is_feature_col`, `count`, and `get_columns`). I know that they are tested through the other, existing tests, but I think they will be more robust to future changes if they themselves are tested.

File: pycytominer/cyto_utils/cells.py
Comment (shntnu) [reply pycytominer/cyto_utils/cells.py:400]: Some quick context here for @bunnech 

The so-called canonical -- or default -- compartments are `cells`, `cytoplasm`, and `nuclei` (and their corresponding column name prefixes are `Cells_`, `Cytoplasm_`, and `Nuclei_`, respectively).

But in general, it's possible to 
- have other ("non-canonical") compartments, and 
- (I think), not have some of these canonical compartments

File: pycytominer/tests/test_cyto_utils/test_cells.py
Comment (bunnech) [reply pycytominer/tests/test_cyto_utils/test_cells.py:312]: In order to read in features faster, I load them into a `np.array` (which I initialize with `dtype=np.float64`). Original metadata might be a `np.int64`, but this should not matter.

File: pycytominer/cyto_utils/cells.py
Comment (bunnech) [reply pycytominer/cyto_utils/cells.py:400]: I compared with `self.compartments`. Thanks for the suggestion @gwaybio  and explanation @shntnu!

File: pycytominer/tests/test_cyto_utils/test_cells.py
Comment (bunnech) [reply pycytominer/tests/test_cyto_utils/test_cells.py:243]: Fixed it by checking whatever is provided in `self.compartments`.

**Comment by bunnech** [comment]
> 2. I am inclined to ask for tests for the three new functions you added (prior to name change, `is_feature_col`, `count`, and `get_columns`). I know that they are tested through the other, existing tests, but I think they will be more robust to future changes if they themselves are tested.

I removed `is_feature_col` now. The other two functions are tested via pytests on `load_compartment`. Not sure which additional test to design.

**Comment by bunnech** [comment]
@gwaybio, I pushed the edits suggested by you. Let me know if anything else is required!

Thanks again!

File: pycytominer/tests/test_cyto_utils/test_cells.py
Comment (gwaybio) [reply pycytominer/tests/test_cyto_utils/test_cells.py:312]: sounds good!

File: pycytominer/cyto_utils/cells.py
Comment (bunnech) [reply pycytominer/cyto_utils/cells.py:455]: Ah, we can equally write:

```
return pd.concat([metas, pd.DataFrame(columns=feat_cols, data=feats)], axis=1)
```

Is this preferred?

Comment (bunnech) [reply pycytominer/cyto_utils/cells.py:455]: We are only using indexing via `:num_meta` or `num_meta:` for the SQL query result.

**Review by gwaybio — COMMENTED** [review]
> > I'm wondering why we need to separate metadata features and morphology features at all. Would it make sense to process all features the same way? Ah, but maybe if we do that they won't be in the correct order in the final sc_df?

> We are required to load the morphology features as a np.array as this is much faster than adding it via a pd.DataFrame. Then I merge both into a joint pd.DataFrame, which I return per compartment. Metadata features might contain strings or objects, so a np.array here is not possible. That's why we separate them.

Got it! Makes perfect sense, thanks for your explanation.

> > I am inclined to ask for tests for the three new functions you added (prior to name change, is_feature_col, count, and get_columns). I know that they are tested through the other, existing tests, but I think they will be more robust to future changes if they themselves are tested.

> I removed is_feature_col now. The other two functions are tested via pytests on load_compartment. Not sure which additional test to design.

I like the elegance of removing `is_feature_col`. The other two functions I think can be tested rather easily:

## `count_sql_table_rows()`

```python
# INSERT TESTING FUNCTION DEFINITION
def test_sc_count_sql_table():  # or something
    # Some python pseudocode testing logic:
    sc = SingleCells()  # Initiate class equivalently as we do in other

    # Iterate over initialized compartments
    for compartment in sc.compartments:
         result_row_count = sc.count_sql_table_rows(table=compartment)
         assert  result_row_count == whatever_example_initialized as
```

## `get_sql_table_col_names()`

```python
# INSERT TESTING FUNCTION DEFINITION
def test_get_sql_table_col_names():  # or something
    # Some python pseudocode testing logic:
    sc = SingleCells()  # Initiate class equivalently as we do in other

    # Iterate over initialized compartments
    for compartment in sc.compartments:
         meta_cols, feat_cols = sc.get_sql_table_col_names(table=compartment)
         assert  meta_cols == expected_meta_cols
         assert  feat_cols == feat_cols
```

File: pycytominer/cyto_utils/cells.py
Comment (gwaybio) [reply pycytominer/cyto_utils/cells.py:455]: The standard is to have metadata features before morphology features in profile data.

>  Ah, we can equally write:
> return pd.concat([metas, pd.DataFrame(columns=feat_cols, data=feats)], axis=1)
> Is this preferred?


If I am correct in my read of the `pd.concat()` call that metadata will be added as the last columns, then yes please.

**Comment by bunnech** [comment]
> The other two functions I think can be tested rather easily:

Thanks! Added some pytests!

File: pycytominer/tests/test_cyto_utils/test_cells.py
Comment (gwaybio) [inline pycytominer/tests/test_cyto_utils/test_cells.py:246]:
```suggestion
def test_sc_count_sql_table():
```

File: pycytominer/tests/test_cyto_utils/test_cells.py
Comment (gwaybio) [inline pycytominer/tests/test_cyto_utils/test_cells.py:253]:
```suggestion
def test_get_sql_table_col_names():
```

Please also assert `feat_cols`

**Review by gwaybio — APPROVED** [review]
Two suggestions, one additional request, and one unresolved comment.

After you address these three things, I will merge (pending passing tests)

File: pycytominer/tests/test_cyto_utils/test_cells.py
Comment (bunnech) [reply pycytominer/tests/test_cyto_utils/test_cells.py:253]: For this I have no consistent ground truth.

Comment (gwaybio) [reply pycytominer/tests/test_cyto_utils/test_cells.py]: won't this (or something close) work?

```python
assert feat_cols == [f"{compartment}_a", f"{compartment}_b", f"{compartment}_c", f"{compartment}_d"]
```

Comment (bunnech) [reply pycytominer/tests/test_cyto_utils/test_cells.py]: Cytoplasm breaks the cycle as it has some additional things. But I could do:

```
assert f"{compartment.capitalize()}_a" in feat_cols
```

**Comment by bunnech** [comment]
> Two suggestions, one additional request, and one unresolved comment.

Addressed them! Thanks a lot for the feedback and guidance!

File: pycytominer/cyto_utils/cells.py
Comment (gwaybio) [reply pycytominer/cyto_utils/cells.py:455]: Just to clarify. My read of `pd.concat()` is incorrect? Metadata features are inserted as the first columns?

Comment (bunnech) [reply pycytominer/cyto_utils/cells.py:455]: No, currently not. Because the existing pytests assume that the structure is morphology features, then meta features.

Comment (gwaybio) [reply pycytominer/cyto_utils/cells.py:455]: Interesting! I added #220 (this fix is beyond scope of your PR)

Comment (bunnech) [reply pycytominer/cyto_utils/cells.py:455]: I changed the existing pytests to agree with the order meta features, then morphological features, and swapped the order of the concat.

Comment (gwaybio) [reply pycytominer/cyto_utils/cells.py:455]: wonderful!

**Comment by gwaybio** [comment]
closes #220 

Thanks again @bunnech !

**Pull request merged**
